### ISSUE TITLE
As a gardener operator, I want to statically assign initial resources (CPU, memory) to Shoot kube-apiservers instead of doing it depending on maximum Node count, such that clusters with little usage aren't wasting that many resources anymore.

### ISSUE THREAD

**Issue by voelzmo** [gardener/gardener#7985]
**How to categorize this issue?**

/area auto-scaling
/area control-plane
/kind enhancement

**Disclaimer: The following descriptions assume you have HVPA enabled. For Seeds where HVPA is disabled, the behavior is different!**

### Problem background

The current scaling mechanism for Shoot kube-apiservers works in three steps
* the kube-apiservers get an initial amount of resources, depending on the `ScalingClass` associated with the Shoot.  See the next section for more details about how this is done and some context why this a was built in the first place.
* HVPA uses HPA's scaling recommendation to scale horizontally until `maxReplicas` is hit and 
* only afterwards uses VPAs scaling recommendation to adjust the kube-apiserver resources

This works reasonably well for clusters, which see a high amount of use and scale to `maxReplicas` and afterwards get more or fewer resources assignes for each replica, but falls short when clusters see little usage and their kube-apiserver are never scaled to more replicas. In this scenario, we end up with kube-apiservers which have requests much higher than they actually need, but VPA recommendations are never applied by HVPA.

### Proposed Solution

We propose to remove the dynamic configuration of initial kube-apiserver resources as it shouldn't be necessary anymore for control-plane stability and replace it with a static set of initial resources. To make the switch from the current implementation towards a statically defined set of initial resources controllable by operators, we propose adding a new feature gate.

### Current implementation details and historical context

The current implementation [sums up the maximum possible Node count over all worker pools](https://github.com/gardener/gardener/blob/0c3a83e464e01991bfbd73514a8aa3a4b3585714/pkg/operation/botanist/kubeapiserver.go#L121-L124) and [assigns one of four `ScalingClass`es based on this](https://github.com/gardener/gardener/blob/0c3a83e464e01991bfbd73514a8aa3a4b3585714/pkg/operation/botanist/kubeapiserver.go#L165-L178). The `ScalingClass` determines the [initial resource requests for the Shoot's kube-apiserver](https://github.com/gardener/gardener/blob/0c3a83e464e01991bfbd73514a8aa3a4b3585714/pkg/operation/botanist/kubeapiserver.go#L180-L207).

This implementation was introduced at a time of "v1 Vertical Pod Autoscaler", which didn't live up to its promise and lacked a lot of the functionality that we now have – that's why it was later re-implemented as "v2 VPA", which is what we're using now. See [the original auto-scaling story](https://github.com/gardener/gardener/issues/16) and [the PR introducing `ScalingClass`](https://github.com/gardener/gardener/issues/255) for more context. We needed a working solution at that time and couldn't wait for "v2 VPA" to be available, because of [concrete issues during operations of clusters under higher load](https://github.com/gardener/gardener/issues/79).

### Implementation details for proposed static assignment of initial resources

* introduce a new feature gate on gardenlet, called `DisableScalingClassForShoots`
* The [current `ScalingClass` implementation](https://github.com/gardener/gardener/blob/0c3a83e464e01991bfbd73514a8aa3a4b3585714/pkg/operation/botanist/kubeapiserver.go#L165-L200) remains intact and gets an additional case for when the above feature gate `DisableScalingClassForShoots` is set to `true`
* The additional case doesn't consider Node count, but just returns a statically configured set of resources, for example
```
requests:
  cpu: 500m
  memory: 1Gi
```

### Rollout/Feature gate removal

The feature gate will be introduced as `alpha`, with the default `false`, in the next release move to `beta` with default `true`, in the following release move to `GA` with default `true` and in the following release removed. We don't expect any special criteria to be necessary for feature gate graduation.

### Relation to GEP-23 and removing HVPA from Shoot kube-apiserver scaling

After [GEP-23](https://github.com/gardener/gardener/pull/7410) has been implemented and HVPA is no longer used to scale Shoot kube-apiservers, we can remove the `ScalingClass` implementation, as we get vertical Pod autoscaling right from the start of a kube-apiserver Pod. The VPA will adjust the resources to the necessary level automatically and we don't have to wait for horizontal scaling to max out. However, GEP-23 as a proposal hasn't been merged yet, therefore the implementation and rollout is still several months from now. With this proposal we want to enable a quick way to reduce the resource waste caused by clusters with little usage and high maximum Node count.

The static assignment of initial resources proposed here will be useful even after GEP-23 has been implemented, such that operators can have different start values for their Shoot control planes e.g. based on if this is a development or testing landscape or a production landscape.

### Risks

* Operators could configure the set of initial resources so low that kube-apiservers fail right after starting the Pod without giving VPA the necessary time and amount of usage samples to make reasonable recommendations. This could lead to control-plane downtimes or perceived longer times until a control-plane is ready after starting a cluster.
* Operators could configure the set of initial resource way too high, resulting in even more resource waste than choosing the initial resources based on the maximum Node count

### Considered Alternatives

Switching with a Seed spec property instead of a feature gate and allowing operators to specify the concrete values for the initial Shoot kube-apiserver resources:

* Add a new property `shootApiserverInitialResources` to [`.spec.settings` on a `Seed`](https://gardener.cloud/docs/gardener/api-reference/core/#seedsettings) of type [`ResourceList`](https://pkg.go.dev/k8s.io/api/core/v1#ResourceList)
* Example content
```
spec:
  settings:
    shootApiserverInitialResources:
      cpu: 500m
      memory: 1Gi
```

We decided against this approach for a number of reasons discussed below.
* we don't want to make the Seed API bigger than necessary, so we're trying to avoid adding new fields if possible
* we don't want to treat kube-apiserver differently than the other control-plane components. Currently, we don't allow any detailed configuration of initial resources, and we think this shouldn't be necessary for most cases. If operators really find they cannot do without this, there are other options, like kyverno or OPA.

**Comment by rfranzke** [comment]
Given that all this scaling-class annotation is marked as 'alpha', we could also remove it immediately w/o further notice. Would this be an option as well? If not, then we should probably find another way to allow operators configuring the initial resources. I don't think it's reasonable to add a new field to the `core` API which basically is deprecated already before initial introduction.

**Comment by voelzmo** [comment]
>  I don't think it's reasonable to add a new field to the core API which basically is deprecated already before initial introduction.

Can you elaborate a bit on this? I'm not sure I understood what you mean. I assume you're saying that the new field added to the Seed would need to be removed once we remove the `ScalingClass` from Gardener code?
Above, I tried to argue that what we're proposing to add will _still_ remain useful, even after `ScalingClass` and HVPA are completely removed:

> The static assignment of initial resources proposed here will be useful even after GEP-22 has been implemented, such that operators can have different start values for their Shoot control planes e.g. based on if this is a development or testing landscape or a production landscape.

Does this make sense?

PS:
> Given that all this scaling-class annotation is marked as 'alpha', we could also remove it immediately w/o further notice. 

I don't think we should do this. This is not about a feature exposed to stakeholders directly (this is not about someone setting this annotation themselves), but about how Gardener determines the initial set of resource requests for Shoot kube-apiservers automagically. This is how we're configuring our landscapes and that's the behavior operators know and stakeholders work with every day. Removing/changing this immediately is something I'd like to avoid, given this risk

> Operators could configure the set of initial resources so low that kube-apiservers fail right after starting the Pod without giving VPA the necessary time and amount of usage samples to make reasonable recommendations. This could lead to control-plane downtimes or perceived longer times until a control-plane is ready after starting a cluster.

Finding the correct value might need some fiddling with the values, and may even be different on DEV and PROD landscapes, so I think we should avoid a solution that requires code changes to adjust the resources or fixes all landscapes on the same set of resources.

> we should probably find another way to allow operators configuring the initial resources

I'm open to suggestions, hence this issue. Which other option do you have in mind?

**Comment by rfranzke** [comment]
> > The static assignment of initial resources proposed here will be useful even after GEP-22 has been implemented, such that operators can have different start values for their Shoot control planes e.g. based on if this is a development or testing landscape or a production landscape.
> 
> Does this make sense?

TBH, I don't think so. We should try to avoid treating the API server pods specially. Otherwise, why wouldn't we expose similar "initial resource" configurations for other control plane components (e.g., etcd or controllers)? 
If this is really needed, perhaps it's a better approach to let operators use third-party tools like Kyverno/OPA if they really want to have such behaviour for testing/development landscapes.

> > Operators could configure the set of initial resources so low that kube-apiservers fail right after starting the Pod without giving VPA the necessary time and amount of usage samples to make reasonable recommendations. This could lead to control-plane downtimes or perceived longer times until a control-plane is ready after starting a cluster. 

> Finding the correct value might need some fiddling with the values, and may even be different on DEV and PROD landscapes, so I think we should avoid a solution that requires code changes to adjust the resources or fixes all landscapes on the same set of resources.

How are such values specific to the landscape purpose like DEV or PROD? How are they specific to all control planes on a seed cluster at all? In my understanding, they are at most specific for dedicated shoots, so I cannot see how a seed-global configuration can be the appropriate way of addressing this problem. Perhaps it's more reasonable to address these downsides in VPA directly?

**Comment by voelzmo** [comment]
> How are such values specific to the landscape purpose like DEV or PROD?

My thinking was that on DEV you might be fine defining any start values much more aggressively than on a production system. I do understand your point that we're not doing this for any other components in Gardener, and it is reasonable to not start doing this for a single component.

> How are they specific to all control planes on a seed cluster at all?

They are not. The only reason why we proposed making this a property on a Seed is to allow for controlled rollout. That's what was meant above with 
> To make the switch from the current implementation towards a statically defined set of resources controllable by operators, we propose that the static resource configuration is not a part of the Gardener code, but a new property on the Seed resource.

Alternatively, we could also have static values in the code and make this a feature gate on gardenlet?

> Perhaps it's more reasonable to address these downsides in VPA directly?

I'm not sure which downsides you're suggesting to address in VPA, could you elaborate on this a bit? My idea with this issue was to get to the point where we can remove code and a band-aid implementation from Gardener code, not make any real changes to autoscaling behavior.

**Comment by rfranzke** [comment]
IIUC, your goal is to drop the initial static resources assignment from the code entirely because "it shouldn't be necessary anymore". In this light, adding a new field to the API that is deprecated before it is even introduced is rather confusing. If you want to make the rollout controllable, then you can indeed just use a feature gate that disables the static resource assignment behaviour, and that eventually gets promoted to GA and removed again. Feature gates are easier to graduate and remove than changing the user-facing APIs.

> I'm not sure which downsides you're suggesting to address in VPA, could you elaborate on this a bit?

I wrote this because you said "we should probably find another way to allow operators configuring the initial resources" - this contradicts with the initial goal description of this issue (removing initial static resources assignments entirely). My point was that we should rather improve the autoscaling in case we are not happy with the scale-up behaviour when new clusters right away get a lot of load (which was the motivation to introduce the static assignments in the beginning).

TL;DR: If you just want to make the removal of the code controllable, let's use a feature gate that we actively promote and drop again within a few release cycles. If you want to allow operators to configure initial resources assignments, let's make them rely on external tools like Kyverno/OPA instead of only exposing this possibility for API servers.

**Comment by voelzmo** [comment]
> IIUC, your goal is to drop the initial static resources assignment from the code entirely (...)

That's the opposite of what's described here. As the title of this issue says, this is **not** about removing a static definition of initial resources, but removing the _dynamic_ logic of doing this according to whatever heuristics and _replace it_ with a static initial set of resources. We still want to define _some_ value for resource requests and not leave this empty.

I'm fine with editing the above to introduce a new feature gate in gardenlet to disable the `ScalingClass` dynamic assignment and have a static assignment for the initial resource requests instead.

**Comment by rfranzke** [comment]
OK, if

> We still want to define *some* value for resource requests and not leave this empty.

does not mean that a human operator still needs to define this somehow via Gardener means (not via third-party tools), then I'm fine. From Gardener PoV, I think we should treat KAPI similar to other components.

**Comment by voelzmo** [comment]
Edited the original post above:
* moved the original proposal of adding a Seed property into "Alternatives"
* changed the proposed implementation to a feature gate

WDYT?

**Comment by rfranzke** [comment]
Thanks, sounds much better now! Except for the "Rollout/Feature gate removal" section:

- the open source repository should not contain any specifics to any special landscape (what is "our own landscape" in the open source context?)
- the feature gate promotion should happen as usual, independent of how landscape operators decide to use or not use it: (1) introduced as alpha, disabled by default. (2) promoted to beta, enabled by default. (3) promoted to GA, locked to default (enabled). (4) removed from the code base. Hence, four release cycles will be needed to cover the full lifecycle.
